helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -344,6 +367,13 @@ private void schedulePendingMessageCleanUp(
             if (accessor.removeProperty(msg.getKey(accessor.keyBuilder(), instanceName))) {
               LogUtil.logInfo(logger, _eventId, String
                   .format("Deleted message %s from instance %s", msg.getMsgId(), instanceName));
+              if (staleMessageMap != null && staleMessageMap.containsKey(msg.getTgtName())
+                  && staleMessageMap.get(msg.getTgtName()).containsKey(msg.getMsgId())) {
```

Comment:
nit, could be simpler by
if (staleMessageMap != null && staleMessageMap.getOrDefault(msg.getTgtName(), EmptyMap()).remove(msg.getMsgId())) {
  if (staleMessageMap.get(msg.getTgtName()).size() == 0) {
                  staleMessageMap.remove(msg.getTgtName());
                }
}

Reply:
Please take a look at the updated one. I believe it's safe.

Reply:
Both versions are safe. Just too verbose.